jvm.sh

```diff
@@ -1,66 +1,104 @@
+
+  # osx jdk strategy
+  test -z "$new" -a -e /usr/libexec/java_home && \
+    new="$(/usr/libexec/java_home -v 1."$version" || true)"
+
+  # sanity check: new must be a folder.
+  test ! -z "$new" -a -d "$new" || return 1
+
+  # PATH cleanup
+  # shellcheck disable=SC2155
+  test ! -z "$previous" -a "$previous" != "$new" && \
+    export PATH="$(echo "$PATH" | sed -e 's|'"$previous"'/bin:||g')"
+
+  # finally export new home and path
+  export JAVA_HOME="$new"
+  export PATH="${JAVA_HOME}/bin:$PATH"
 }
 
-# shellcheck disable=SC2039
-_jvm-discover-version() {
-  version=""
-  if [ -f pom.xml ]; then
-    version="$(\
-      grep '<java.version>' pom.xml | \
-      sed 's/.*<java.version>1.\(.*\)<\/java.version>.*/\1/' \
-    )"
-  fi
-  if [ -z "$version" ] && [ -f .java-version ]; then
-    version="$(cat .java-version)"
-  fi
-  if [ -z "$version" ] && [ -f ~/.java-version ]; then
-    version="$(cat ~/.java-version)"
-  fi
+# evaluates the 'maven.compiler.source' expression, returning the found java
+# version
+__jvm_pomversion_evaluate() {
+  MAVEN_OPTS="" mvn help:evaluate \
+    -Dexpression='maven.compiler.source' |
+    grep -e '^1\.[4-9]$' |
+    cut -f2 -d'.'
+}
+
+# tried to find the java version using regex.
+__jvm_pomversion_regex() {
+  grep -Eo '<(java.version|maven.compiler.source|source)>1\.[4-9]</.*>' pom.xml |
+    cut -f2 -d'>' |
+    cut -f2 -d'.' |
+    cut -f1 -d'<'
+}
+
+# tries multiple strategies to find the java version, and then sets it in a
+# .java-version
+__jvm_pomversion() {
+  test -f pom.xml || return 1
+  version="$(__jvm_pomversion_regex)"
+  test -z "$version" && version="$(__jvm_pomversion_evaluate)"
+  test ! -z "$version" && echo "$version" > .java-version
+}
+
+# finds out which java version should be used.
+__jvm_version() {
+  test ! -f .java-version && __jvm_pomversion
+  test -f .java-version && version="$(cat .java-version)"
+  test -z "$version" -a -f ~/.java-version && version="$(cat ~/.java-version)"
   echo "$version"
 }
 
-# shellcheck disable=SC2039
-_jvm-discover-and-set-version() {
-  version="$(_jvm-discover-version)"
-  [ ! -z "$version" ] && _jvm_set-java-path "$version"
+# called when a dir changes. Find which java version to use and sets it to PATH.
+__jvm_main() {
+  version="$(__jvm_version)"
+  test ! -z "$version" && __jvm_set "$version"
 }
 
-# shellcheck disable=SC2039
-_jvm-edit-config() {
-  if [ ! -f ~/.jvmconfig ]; then
-    cat > ~/.jvmconfig <<EOF
-custom-jdk-7=Path to custom jdk 7
-custom-jdk-8=Path to custom jdk 8
-EOF
-  fi
+# edit custom java version configurations
+__jvm_config() {
+  test ! -f ~/.jvmconfig && echo "custom-jdk=/path/to/custom/jdk" > ~/.jvmconfig
   $EDITOR ~/.jvmconfig
 }
 
-# shellcheck disable=SC2039
-_jvm-command-list() {
-  echo "local global version reload config"
+# shows usage instructions
+__jvm_usage() {
+  cat <<EOF
+NAME:
+  jvm - The Java Version Manager
+
+USAGE:
+  jvm command [command options]
```

Comment:
I find it unusual to put two arguments like that in the usage. In general, hyphens are used to keep it a single argument:

```
jvm command [command-options]
```

Some examples:

```
sed [OPTION]... {script-only-if-no-other-script} [input-file]...
mawk [-W option] [-F value] [-v var=value] [-f program-file] [--] [file ...]
chmod [OPTION]... OCTAL-MODE FILE...
```